openmc/deplete/pool.py

```diff
@@ -59,6 +64,52 @@ def deplete(func, chain, x, rates, dt, matrix_func=None):
     else:
         matrices = map(matrix_func, repeat(chain), rates, fission_yields)
 
+    if msr is not None:
+        # Calculate removal rate terms as diagonal matrices
+        removals = map(chain.form_rr_term, repeat(msr), msr.burnable_mats)
+        # Subtract removal rate terms from Bateman matrices
+        matrices = [matrix - removal for (matrix, removal) in zip(matrices, removals)]
+
+        if len(msr.index_transfer) > 0:
+            # Calculate transfer rate terms as diagonal matrices
+            transfers = list(map(chain.form_rr_term, repeat(msr),
+                                msr.index_transfer))
+
+            # Combine all matrices together in a single matrix of matrices
+            # to be solved in one-go
+            n_rows = n_cols = len(msr.burnable_mats)
+            rows = []
+            rows_cols = \
+                np.array(np.meshgrid(range(n_rows),
+                                     range(n_cols))).T.reshape(-1, 2)
+            last_col = n_cols - 1
+            cols = []
+            for (row, col) in rows_cols:
```

Comment:
I would opt for a simple nested for-loop rather than the convoluted use of meshgrid and array reshaping (which makes it hard for a reader to figure out what's going on):
```Python
for row in range(n_rows):
    for col in range(n_cols):
```